Freshness Edit: dotnet content (#45187)
dotnet/docs: docs/csharp/asynchronous-programming/index.md

---
title: Asynchronous programming
description: An overview of the C# language support for asynchronous programming using async, await, Task, and Task<T>
ms.date: 09/28/2023
description: Explore an overview of the C# language support for asynchronous programming by using async, await, Task, and Task<T>.
ms.date: 03/10/2025
---
# Asynchronous programming with async and await

> The [Task asynchronous programming model (TAP)](task-asynchronous-programming-model.md) provides an abstraction over asynchronous code. You write code as a sequence of statements, just like always. You can read that code as though each statement completes before the next begins. The compiler performs many transformations because some of those statements may start work and return a <xref:System.Threading.Tasks.Task> that represents the ongoing work.

Reviewer: I think TAP is a fairly well-known acronym for this type of async programming, so might be good to keep it.

Author: The TAP acronym is already applied to the pattern. I thought it would be confusing to reuse it here in a different way.

> The [task asynchronous programming model](task-asynchronous-programming-model.md) provides a layer of abstraction over typical asynchronous coding. In this model, you write code as a sequence of statements, the same as usual. The difference is you can read your task-based code as the compiler processes each statement and before it starts processing the next statement. To accomplish this model, the compiler performs many transformations to complete each task. Some statements can initiate work and return a <xref:System.Threading.Tasks.Task> object that represents the ongoing work and the complier must resolve these transformations. The goal of task asynchronous programming is to enable code that reads like a sequence of statements, but executes in a more complicated order. Execution is based on external resource allocation and when tasks complete.

Reviewer: ```suggestion
The [task asynchronous programming model](task-asynchronous-programming-model.md) provides a layer of abstraction over typical asynchronous coding. In this model, you write code as a sequence of statements, the same as usual. The difference is you can read your task-based code as the compiler processes each statement and before it starts processing the next statement. To accomplish this model, the compiler performs many transformations to complete each task. Some statements can initiate work and return a <xref:System.Threading.Tasks.Task> object that represents the ongoing work and the compiler must resolve these transformations. The goal of task asynchronous programming is to enable code that reads like a sequence of statements, but executes in a more complicated order. Execution is based on external resource allocation and when tasks complete.
```

1. Pour a cup of coffee.
1. Heat a pan, then fry two eggs.
1. Fry three slices of bacon.
1. Toast two pieces of bread.
1. Spread butter and jam on the toast.
1. Pour a glass of orange juice.

If you have experience with cooking, you might complete these instructions **asynchronously**. You start warming the pan for eggs, then start frying the bacon. You put the bread in the toaster, then start cooking the eggs. At each step of the process, you start a task, and then transition to other tasks that are ready for your attention.

Cooking breakfast is a good example of asynchronous work that isn't parallel. One person (or thread) can handle all the tasks. One person can make breakfast asynchronously by starting the next task before the previous task completes. Each cooking task progresses regardless of whether someone is actively watching the process. As soon as you start warming the pan for the eggs, you can begin frying the bacon. After the bacon starts to cook, you can put the bread in the toaster.

> For a parallel algorithm, you need multiple people who cook (or multiple threads). One person cooks the eggs, another fries the bacon, and so on. Each person focuses on their one specific task. Each person who is cooking (or thread) is blocked synchronously awaiting for the current task to complete: Bacon ready to flip, Bread ready to pop up in toaster, and so on.

Reviewer: ```suggestion
For a parallel algorithm, you need multiple people who cook (or multiple threads). One person cooks the eggs, another fries the bacon, and so on. Each person focuses on their one specific task. Each person who is cooking (or each thread) is blocked synchronously waiting for the current task to complete: Bacon ready to flip, bread ready to pop up in toaster, and so on.
```

:::image type="content" source="media/synchronous-breakfast.png" border="false" alt-text="Diagram that shows instructions for preparing breakfast as a list of seven sequential tasks completed in 30 minutes.":::

Consider the same list of synchronous instructions written as C# code statements:

:::code language="csharp" source="snippets/index/AsyncBreakfast-starter/Program.cs" highlight="15-34":::

> When you cook the breakfast in a synchronous manner, it takes about 30 minutes to prepare. The duration is the sum of the individual task times. But computers don't interpret cooking instructions the same way people do. The computer blocks each task statement until all work completes, and then it proceeds to the next task statement. This approach can take significant time. In the breakfast example, the computer method creates an unsatisfying breakfast. Later tasks in the synchronous list like toasting the bread don't start until earlier tasks complete. Some food gets cold before the breakfast is ready to serve.

Reviewer: ```suggestion
When you cook the breakfast in a synchronous manner, it takes about 30 minutes to prepare. The duration is the sum of the individual task times. But computers don't interpret cooking instructions the same way people do. The computer blocks each task statement until all work completes, and then it proceeds to the next task statement. This approach can take significant time. In the breakfast example, the computer method creates an unsatisfying breakfast. Later tasks in the synchronous list, like toasting the bread, don't start until earlier tasks complete. Some food gets cold before the breakfast is ready to serve.
```

> If you want the computer to execute instructions asynchronously, you must write asynchronous code. When you write client programs, you want the UI to be responsive to user input. Your application shouldn't freeze all interaction while downloading data from the web. When you write server programs, you don't want to block threads that might be serving other requests. Using synchronous code when asynchronous alternatives exist, hurts your ability to scale out less expensively. You pay for blocked threads.

Reviewer: ```suggestion
If you want the computer to execute instructions asynchronously, you must write asynchronous code. When you write client programs, you want the UI to be responsive to user input. Your application shouldn't freeze all interaction while downloading data from the web. When you write server programs, you don't want to block threads that might be serving other requests. Using synchronous code when asynchronous alternatives exist hurts your ability to scale out less expensively. You pay for blocked threads.
```

Reviewer: I don't think that comma really belongs there.

Successful modern apps require asynchronous code. Without language support, writing asynchronous code requires callbacks, completion events, or other means that obscure the original intent of the code. The advantage of synchronous code is the step-by-step action that makes it easy to scan and understand. Traditional asynchronous models force you to focus on the asynchronous nature of the code, not on the fundamental actions of the code.

## Don't block, await instead

The previous code highlights an unfortunate programming practice: Writing synchronous code to perform asynchronous operations. The code blocks the current thread from doing any other work. The code doesn't interrupt the thread while there are running tasks. The outcome of this model is similar to you staring at the toaster after you put in the bread. In this scenario, you ignore any interruptions and don't start other tasks until the bread pops up. You might ignore a phone call, forget to take the butter and jam out of the fridge, or miss seeing a fire starting on the stove.

> In your daily life, you don't want to overlook problems or miss opportunities. You want to both toast the bread and handle other concerns at the same time. The same is true with your code. 

Reviewer: I'm not sure this is the best analogy - I think it's more about being "uninterruptible". But I'll leave it to Bill to decide.

Author: I removed the extra statement and adjusted the phrasing.

> [!NOTE]
> The updated code doesn't yet take advantage of key features of asynchronous programming, which can result in shorter completion times. The code processes the tasks in roughly the same amount of time as the initial synchronous version. For the full method implementations, see the [final version](#review-final-version) later in this article. 

Let's apply the breakfast example to the updated code. The thread doesn't block while the eggs or bacon are cooking, but the code also doesn't start other tasks until the current work completes. You still put the bread in the toaster and stare at the toaster until the bread pops up, but you can now respond to interruptions. In a restaurant where multiple orders are placed, the cook can start a new order while another is already cooking.

> In the updated code, the thread working on the breakfast isn't blocked while awaiting for any started task that's unfinished. For some applications, this change is all you need. You can enable your app to support user interaction while data downloads from the web. But in other scenarios, you want more. You don't want each component task to execute sequentially. A better approach is to start each component task before awaiting for the previous task to complete.

Reviewer: ```suggestion
In the updated code, the thread working on the breakfast isn't blocked while waiting for any started task that's unfinished. For some applications, this change is all you need. You can enable your app to support user interaction while data downloads from the web. But in other scenarios, you want more. You don't want each component task to execute sequentially. A better approach is to start each component task while waiting for the previous task to complete.
```

## Start tasks concurrently

For most operations, you want to start several independent tasks immediately. As each task completes, you initiate other work that's ready to start. When you apply this methodology to the breakfast example, you can prepare breakfast more quickly. You also get everything ready close to the same time, so you can enjoy a hot breakfast.

> The <xref:System.Threading.Tasks.Task?displayProperty=nameWithType> class and related types are classes you can use to apply this style of reasoning to tasks that are in progress. This approach enables you to write code that more closely resembles the way you create breakfast in real life. You start cooking the eggs, bacon, and toast at the same time. As each food item requires action, you turn your attention to that task, take care of the action, and then await for something else that requires your attention.

Reviewer: ```suggestion
The <xref:System.Threading.Tasks.Task?displayProperty=nameWithType> class and related types are classes you can use to apply this style of reasoning to tasks that are in progress. This approach enables you to write code that more closely resembles the way you create breakfast in real life. You start cooking the eggs, bacon, and toast at the same time. As each food item requires action, you turn your attention to that task, take care of the action, and then wait for something else that requires your attention.
```

Reviewer: "await for" sounds weird to me - is that actually correct English?

Author: I agree "await" sounds odd. I standardized on using this term (and forms) to see if the authors desired language that connects to the API keyword, or if "wait" is sufficient.

Author: I adjusted the phrasing throughout. In most cases, I replaced "await" with "wait" (or the appropriate form).
